Source/ARTRealtimePresence.m

```diff
@@ -622,4 +660,263 @@ - (void)failPendingPresence:(ARTStatus *)status {
     }
 }
 
+- (void)broadcast:(ARTPresenceMessage *)pm {
+    [_eventEmitter emit:[ARTEvent newWithPresenceAction:pm.action] with:pm];
+}
+
+- (void)onAttached:(ARTProtocolMessage *)message {
+    [self startSync];
+    if (!message.hasPresence) {
+        // RTP1 - when an ATTACHED message is received without a HAS_PRESENCE flag, reset PresenceMap
+        [self endSync];
+        ARTLogDebug(self.logger, @"R:%p C:%p (%@) PresenceMap has been reset", _realtime, self, _channel.name);
+    }
+    [self sendPendingPresence];
+    [self reenterInternalMembers]; // RTP17i
+}
+
+- (void)onMessage:(ARTProtocolMessage *)message {
+    int i = 0;
+    for (ARTPresenceMessage *p in message.presence) {
+        ARTPresenceMessage *presence = p;
+        if (presence.data && _dataEncoder) {
+            NSError *decodeError = nil;
+            presence = [p decodeWithEncoder:_dataEncoder error:&decodeError];
+            if (decodeError != nil) {
+                ARTErrorInfo *errorInfo = [ARTErrorInfo wrap:[ARTErrorInfo createWithCode:ARTErrorUnableToDecodeMessage message:decodeError.localizedFailureReason] prepend:@"Failed to decode data: "];
+                ARTLogError(self.logger, @"RT:%p C:%p (%@) %@", _realtime, _channel, _channel.name, errorInfo.message);
+            }
+        }
+
+        if (!presence.timestamp) {
+            presence.timestamp = message.timestamp;
+        }
+
+        if (!presence.id) {
+            presence.id = [NSString stringWithFormat:@"%@:%d", message.id, i];
+        }
+
+        if ([self add:presence]) {
+            [self broadcast:presence];
+        }
+
+        ++i;
+    }
+}
+
+- (void)onSync:(ARTProtocolMessage *)message {
+    _syncMsgSerial = [message.msgSerial longLongValue];
+    _syncChannelSerial = message.channelSerial;
+
+    if (!self.syncInProgress) {
+        [self startSync];
+    }
+    else {
+        ARTLogDebug(self.logger, @"RT:%p C:%p (%@) PresenceMap sync is in progress", _realtime, _channel, _channel.name);
+    }
+
+    [self onMessage:message];
+
+    if ([_channel isLastChannelSerial:message.channelSerial]) {
```

Comment:
Can you move `isLastChannelSerial` method to `Source/ARTRealtimePresence.m`. It doesn't make sense to keep it as a part of `realtimechannel.m` file.

Reply:
Also, `isLastChannelSerial` should be redefined as
```
- (bool)isLastChannelSerial:(NSString *)channelSerial {
    if ([channelSerial sEqualToString:@""]) {
        return true;
    }
    NSArray * a = [channelSerial componentsSeparatedByString:@":"];
    if([a count] >1 && ![[a objectAtIndex:1] isEqualToString:@""] ) {
        return false;
    }
    return true;
}
```

Reply:
This is as per spec RTP18b, RTP18c

Reply:
You should check https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L286

Reply:
https://github.com/ably/ably-cocoa/pull/1870/commits/51a00d2dcaef813c06ac9cfe6d3c005285a70ed7